### DISCUSSION TITLE
More than 2 nested x-for breaks order of iteration

### DISCUSSION THREAD

**Discussion by RonMordoch** [alpinejs/alpine#3475] [7. Bugs (archived)] [answered]
I will preface this by a working example of 2 nested `x-for`s :
```
<script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.12.0/dist/cdn.min.js"></script>
<script>
function data(){
  return {
    key_to_values: {
      1: [10, 11],
      2: [20, 21, 22],
      3: [30, 31, 32, 33]
    }          
	};
}
</script>

<div x-data="data()">
  <table>
    <thead>
      <th>Key</th>
      <th>Value</th>
    </thead>
    <tbody>
      <template x-for="k in Object.keys(key_to_values)" :key="k">
        <template x-for="(v, v_idx) in key_to_values[k]" :key="v_idx">
          <tr>
            <td x-text="k"></td>
            <td x-text="v"></td>
          </tr>
        </template>
      </template>
    </tbody>
  </table>
</div>
```

This will correctly create a table, where the "value" column has increasing order of numbers - as it should be over the arrays. The keys when I ran decreased from 3 to 1, but as we are iterating over an object, no order is guaranteed anyway.

Now assume I want to add another layer of looping, an additional `x-for`. In this example, say  I want to duplicate the number of times each value is displayed (3 in this case):
```
<script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.12.0/dist/cdn.min.js"></script>
<script>
function data(){
  return {
    key_to_values: {
      1: [10, 11],
      2: [20, 21, 22],
      3: [30, 31, 32, 33]
    }          
	};
}
</script>

<div x-data="data()">
  <table>
    <thead>
      <th>Key</th>
      <th>Value</th>
    </thead>
    <tbody>
      <template x-for="k in Object.keys(key_to_values)" :key="k">
        <template x-for="(v, v_idx) in key_to_values[k]" :key="v_idx">
        <template x-for="i in 3">          
          <tr>
            <td x-text="k"></td>
            <td x-text="v"></td>
          </tr>
        </template>
      </template>
      </template>
    </tbody>
  </table>
</div>
```
Now immediately, the order of the value column is reversed, and the number are decreasing, unlike the previous example.
This inconsistency affects lots of issues- let say I want to merge all the "key" cells for a single key. So , working with HTML tables, I check if we are in the first element of the array, and only then writes the table cell with the matching rowspan:
```
<script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.12.0/dist/cdn.min.js"></script>
<script>
function data(){
  return {
    key_to_values: {
      1: [10, 11],
      2: [20, 21, 22],
      3: [30, 31, 32, 33]
    }          
	};
}
</script>

<div x-data="data()">
  <table>
    <thead>
      <th>Key</th>
      <th>Value</th>
    </thead>
    <tbody>
      <template x-for="k in Object.keys(key_to_values)" :key="k">
        <template x-for="(v, v_idx) in key_to_values[k]" :key="v_idx">
        <template x-for="i in 3">          
          <tr>
          <template x-if="v_idx == 0">
            <td x-text="k" :rowspan="3* key_to_values[k].length"></td>
          </template>
            <td x-text="v"></td>
          </tr>
        </template>
      </template>
      </template>
    </tbody>
  </table>
</div>
```
This table breaks completely.

I have seen other issues referencing problems with `x-for`'s order, they were marked as solved but this issue persist.
Please also notice I am using the latest version of AlpineJS.

**Comment by ekwoka** [comment, marked as answer]
You need to have a wrapper element inside of the x-for before the next x-for.

There isn't really a good way (maybe there is but it's not immediately clear) for an x-for to know where it's next loop should go other than look at where the last one went and putting it right after it.

Here you're doing nested loops that output a flat nodelist. This is just not supported by alpine.

so you need to ensure the template has a single child that is a real element.

If you need them flat for css purposes, you can use `display: contents;` to have them be passthrough elements for the flow.
